File: virtspawn-core/src/libvirt/node.rs
use virt::connect::Connect;

use crate::state::NodeInfo;
use crate::LibvirtError;

pub fn get_node_info(conn: &Connect) -> Result<NodeInfo, LibvirtError> {
    let hostname = conn
        .get_hostname()
        .map_err(|e| LibvirtError::Operation(format!("Failed to get hostname: {e}")))?;

    let hv_type = conn
        .get_type()
        .map_err(|e| LibvirtError::Operation(format!("Failed to get hypervisor type: {e}")))?;

    let hv_version = conn
        .get_hyp_version()
        .map(|v| format!("{}.{}.{}", v / 1_000_000, (v / 1_000) % 1_000, v % 1_000))
        .unwrap_or_else(|_| "unknown".to_string());

    let lib_version = conn
        .get_lib_version()
        .map(|v| format!("{}.{}.{}", v / 1_000_000, (v / 1_000) % 1_000, v % 1_000))
        .unwrap_or_else(|_| "unknown".to_string());

    let node = conn
        .get_node_info()
        .map_err(|e| LibvirtError::Operation(format!("Failed to get node info: {e}")))?;

    let active_domains = conn.num_of_domains().unwrap_or(0) as u32;
    let defined_domains = conn.num_of_defined_domains().unwrap_or(0) as u32;

    Ok(NodeInfo {
        hostname,
        hypervisor: hv_type,
        hypervisor_version: hv_version,
        lib_version,
        cpu_model: node.model,
        cpu_cores: node.cores,
        cpu_threads: node.threads,
        cpu_sockets: node.sockets,
        memory_mb: node.memory / 1024,
        numa_nodes: node.nodes,
        active_vms: active_domains,
        defined_vms: defined_domains,
    })
}

File: web/src/api/node.ts
// Copyright (c) 2026 ZyvorAI Labs Private Limited. All rights reserved.
// Proprietary software — see LICENSE in the repository root.
// https://zyvor.dev · info@zyvor.dev

import { readJsonObject } from './client'

const API = '/api/v1'

export interface NodeInfo {
  hostname: string
  hypervisor: string
  hypervisor_version: string
  lib_version: string
  cpu_model: string
  cpu_cores: number
  cpu_threads: number
  cpu_sockets: number
  memory_mb: number
  numa_nodes: number
  active_vms: number
  defined_vms: number
}

export interface HealthStatus {
  status: string
  libvirt: boolean
}

export const getNodeInfo = () => readJsonObject<NodeInfo>(`${API}/node`)
export const getHealth = () => readJsonObject<HealthStatus>(`${API}/health`)
